File: javascript/core/src/scope.js
// scope is determined in compile time, not runtime
// This code demonstrates the concept of scope in JavaScript, including function scope, block scope, and the use of eval and with statements.

// Function Scope: variables declared with 'var' are function-scoped, meaning they are accessible within the function they are declared in, but not outside of it.
function greet() {
  var message = "Hello!";
  console.log(message); // "Hello!"
}
// console.log(message); // ReferenceError

if (true) {
  var a = 5;
}
console.log(a); // 5 — leaked out of the block!

// Block Scope: variables declared with 'let' or 'const' are block-scoped, meaning they are only accessible within the block they are defined in
if (true) {
  let b = 10;
  const c = 20;
}
// console.log(b); // ReferenceError: b is not defined
// console.log(c); // ReferenceError: c is not defined



// two exceptions: eval and with
/* eval: allows you to execute a string of JavaScript code in the current scope. due to its dynamic nature, it can lead to security issues and performance problems, so it's generally discouraged in modern JavaScript.
   with: allows you to extend the scope chain for a block of code, but it's generally discouraged due to performance and readability issues */

let x = 10;
eval("x = x + 5");
console.log(x); // 15

// eval("let y = 20; console.log(y);"); // This will throw an error because 'let' is block-scoped and not accessible here
eval("var z = 30; console.log(z);"); // This will work because 'var' is function-scoped and accessible here
console.log(z); // 30

// eval("const a = 40; console.log(a);"); // This will throw an error because 'const' is block-scoped and not accessible here
eval("function test() { return 'Hello from eval'; } console.log(test());"); // This will work and log the message

const person = { name: "Alice", age: 30 };

with (person) {
  console.log(name); // "Alice"
  console.log(age);  // 30
}

File: perl/core/bin/scope.pl
#!/usr/bin/perl
use strict;
use warnings;

# Scoped Variable Declarations
# This script demonstrates variable scope in Perl.
# Variables declared with 'my' are lexically scoped, meaning they are only visible within the block they are defined in.
my $x = 10;
if ($x > 5) {
    my $y = 99;
    print "$x and $y\n";  # both are visible here
}
print "$x\n";     # OK
# print "$y\n";   # ERROR: $y is not visible outside block


# Variables declared with 'our' are package-scoped, meaning they are visible throughout the package.
# They can be accessed from anywhere in the package, including from other files that use the package.
package MyApp;
use strict;
use warnings;

our $version = "1.0";  # accessible as $MyApp::version

package main;
print $MyApp::version;  # works


# This script demonstrates the use of 'local' to temporarily change the value of a variable.
# 'local' creates a temporary value for the variable that is visible in the current block and any nested blocks.
# When the block ends, the original value is restored.
our $message = "Hello";

sub speak {
    print "$message\n";
}

sub shout {
    local $message = "HEY!";
    speak();  # prints "HEY!"
}

speak();  # prints "Hello"
shout();
speak();  # prints "Hello" again


# This script demonstrates the use of 'state' variables in Perl.
# 'state' variables retain their value between calls to the subroutine, but are not visible outside the subroutine.
# They are initialized only once, the first time the subroutine is called.
# This is useful for maintaining state without using global variables.
use feature 'state';
use strict;
use warnings;

sub counter {
    state $count = 0;
    $count++;
    print "Count: $count\n";
}

counter();  # Count: 1
counter();  # Count: 2
counter();  # Count: 3
